File: hs/invitation-service-web/src/main/java/com/hubspot/invitationserviceweb/service/AvailabilityCalculator.java
package com.hubspot.invitationserviceweb.service;

import com.hubspot.invitationserviceweb.model.Country;
import com.hubspot.invitationserviceweb.model.Partner;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.springframework.stereotype.Component;

import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;

@Component
@Slf4j
public class AvailabilityCalculator {

    Map<String,Map<LocalDate,List<String>>> countriesDatesEmails = new HashMap<>();

    public AvailabilityCalculator() { }

    public void processAvailabilities(List<Partner> partners){
        log.info("Loading availability");
        partners.forEach(this::processAvailability);
    }

    public List<Country> processCountries(){
        List<Country> countries = new LinkedList<>();
        for (String country: countriesDatesEmails.keySet()){
            val dto = new Country();
            dto.setName(country);
            List<String> attendeeList = null;
            val dateParticipants = countriesDatesEmails.get(country);
            val selectedDates =
                    dateParticipants.keySet().stream().sorted(Comparator.comparing(date -> -dateParticipants.get(date).size())).collect(Collectors.toList());
            if (selectedDates.isEmpty()){
                dto.setStartDate(null);
            } else {
                LocalDate selectedDate = selectedDates.get(0);
                dto.setStartDate(selectedDate.toString());
                attendeeList = dateParticipants.get(selectedDate);
            }
            if (attendeeList==null){
                dto.setAttendees(new LinkedList<>());
                dto.setAttendeeCount(0);
            } else {
                dto.setAttendees(attendeeList);
                dto.setAttendeeCount(attendeeList.size());
            }
            countries.add(dto);
        }
        return countries;
    }


    public void processAvailability(Partner partner){
        val country = partner.country;
        val dates  = partner.availableDates.stream().map(this::extractDate).collect(Collectors.toList());
        for (int i=0;i<dates.size()-1;i++){
            val date = dates.get(i);
            val nextDate = dates.get(i+1);
            if (date.plusDays(1).equals(nextDate)){
                val countryMap = countriesDatesEmails.get(country);
                if (countryMap==null){
                    //create countryMap
                    log.info("Creating map for country {}",country);
                    Map<LocalDate,List<String>> datesEmails = new TreeMap<>();
                    //Create emailList for that date
                    List<String> emails = new LinkedList<>();
                    emails.add(partner.email);
                    //Add date to that map
                    datesEmails.put(date,emails);
                    countriesDatesEmails.put(country,datesEmails);
                    log.info("Created date {} for new country {}",date,country);
                    log.info("Added participant {} to new date {} on new country {}",partner.email,date,country);
                } else {
                    //update country map
                    val existingDateList = countryMap.get(date);
                    if (existingDateList==null){
                        List<String> emails = new LinkedList<>();
                        emails.add(partner.email);
                        countryMap.put(date,emails);
                        log.info("Created date {} for existing country {}",date,country);
                        log.info("Added participant {} to new date {} on existing country {}",partner.email,date,
                                country);
                    } else {
                        //update date list
                        existingDateList.add(partner.email);
                        log.info("Added participant {} to existing date {} on existing country {}",partner.email,date,
                                country);
                    }
                }
            }
        }
        //Add country in case no match
        if (countriesDatesEmails.get(country)==null){
            Map<LocalDate,List<String>> datesEmails = new TreeMap<>();
            countriesDatesEmails.put(country,datesEmails);
        }
    }


    private LocalDate extractDate(String date) {
        return LocalDate.parse(date);
    }


}

File: hs/akka-invitation-service/src/main/scala/com/hs/invitation/service/service/AvailabilityCalculator.scala
package com.hs.invitation.service.service

import java.time.LocalDate

import com.hs.invitation.service.model.{Country, Partner}

import scala.collection.mutable.{HashMap, Map, TreeMap}
import org.slf4j.LoggerFactory

class AvailabilityCalculator {

  val logger = LoggerFactory.getLogger(getClass.getSimpleName)

  var countriesDatesEMailsMap : Map[String,Map[LocalDate,Seq[String]]] = HashMap.empty[String,Map[LocalDate,Seq[String]]]

  def processAvailability(partner: Partner): Unit = {
    val country = partner.country
    val availableDates = partner.availableDates.map(date => LocalDate.parse(date))
    for (i <- 0 to availableDates.size - 2) {
      val date: LocalDate = availableDates.lift(i).get
      val nextDate = availableDates.lift(i + 1).get
      // Use case that we want to add date
      if (date.plusDays(1).equals(nextDate)) {
        val countryDateEmailMap = countriesDatesEMailsMap.get(country)
        countryDateEmailMap match {
          // country has never been added, but a match exist
          case None =>
            val emails: Seq[String] = Seq(partner.email)
            val newCountryDateEmailMap: Map[LocalDate, Seq[String]] = TreeMap(date -> emails)
            countriesDatesEMailsMap.put(country, newCountryDateEmailMap)
            logger.info(s"Added participant ${partner.email} to newCountry $country on new date $date")
          // country has been added and a match exist
          case Some(existingCountryDateEmailMap) => {
            existingCountryDateEmailMap.get(date) match {
              // country exists, but List for that date does not exist
              case None => {
                val emails: Seq[String] = Seq(partner.email)
                existingCountryDateEmailMap.put(date, emails)
                logger.info(s"Added participant ${partner.email} to existing Country $country on new date $date")
              }
              //country exists, List for that date exist
              case Some(existingEmailListForDate) => {
                val emails : Seq[String] = existingEmailListForDate ++ Seq(partner.email)
                existingCountryDateEmailMap.put(date, emails)
                logger.info(s"Added participant ${partner.email} to existing Country $country on existing date $date")
              }
            }
          }
        }
      }
    }

    countriesDatesEMailsMap.get(country) match {
      case None => {
        val newCountryDateEmailMap: Map[LocalDate, Seq[String]] = TreeMap.empty
        countriesDatesEMailsMap.put(country, newCountryDateEmailMap)
        logger.info(s"Added new empty country ${country}")
      }
      case Some(_) =>
    }

  }


  def processCountries: List[Country] = {
    var countryList = Seq.empty[Country]
    for (country <- countriesDatesEMailsMap.keySet) {
      logger.info(s"Processing country $country")
      val countryName = country
      val dateMapForCountry = countriesDatesEMailsMap.get(country).get
      for (dateEvaluated <- dateMapForCountry.keySet){
        val participantsSize = dateMapForCountry.get(dateEvaluated).get.size
        logger.info(s"For country $countryName on $dateEvaluated there are $participantsSize interested participants")
      }
      val helper1 = dateMapForCountry.keySet.toSeq.sortWith((d1, d2) => dateMapForCountry.get(d1).get.size > dateMapForCountry.get(d2).get.size)
      val maybeSelectedDate= helper1.toList.lift(0)

      maybeSelectedDate match {
        case Some(selectedDate) => {
          val attendeesList: List[String] = dateMapForCountry.get(selectedDate).get.toList
          val attendeeCount = attendeesList.size
          val countryDto = Country(attendeeCount, attendeesList, countryName, selectedDate.toString)
          logger.info(s"Adding country $countryDto")
          countryList = countryList :+ countryDto

        }
        case None => {
          val attendeesList: List[String] = List.empty
          val attendeeCount = 0
          val countryDto = Country(attendeeCount, attendeesList, countryName, null)
          logger.info(s"Adding empty country $countryDto")
          countryList = countryList :+ countryDto
        }
      }
    }
    countryList.toList
  }

}
